node/task.ts

```diff
@@ -243,41 +248,110 @@ export function loc(key: string, ...param: any[]): string {
 // Input Helpers
 //-----------------------------------------------------
 /**
- * Gets a variables value which is defined on the build definition or set at runtime.
+ * Gets a variable value that is defined on the build definition or set at runtime.
  * 
  * @param     name     name of the variable to get
  * @returns   string
  */
 export function getVariable(name: string): string {
     var varval;
-    if (vault) {
-        varval = vault.retrieveSecret('SECRET_' + name.replace(/\./g, '_').toUpperCase());
+
+    // get the metadata
+    var info: VariableInfo;
+    var key: string = getVariableKey(name);
+    if (variableInfo.hasOwnProperty(key)) {
+        info = variableInfo[key];
     }
 
-    if (!varval) {
-        varval = process.env[name.replace(/\./g, '_').toUpperCase()];
+    if (info && info.secret) {
+        // get the secret value
+        varval = vault.retrieveSecret('SECRET_' + key);
+    } else {
+        // get the public value
+        varval = process.env[key];
+
+        // fallback for pre 2.104.1 agent
+        if (!varval && name.toUpperCase() == 'AGENT.JOBSTATUS') {
+            varval = process.env['agent.jobstatus'];
+        }
     }
 
     debug(name + '=' + varval);
     return varval;
 }
 
 /**
- * Sets a variables which will be available to subsequent tasks as well.
+ * Gets all variables that are defined on the build definition or set at runtime.
+ * Requires a 2.104.1 agent or higher.
+ * 
+ * @returns VariableInfo[]
+ */
+export function getVariables(): VariableInfo[] {
+    return Object.keys(variableInfo)
+        .filter(function (key: string) {
+            return variableInfo.hasOwnProperty(key);
+        })
+        .map(function (key: string) {
+            var info: VariableInfo = variableInfo[key];
+            return new VariableInfo(info.name, getVariable(info.name), info.secret);
+        });
+}
+
+/**
+ * Sets a variable which will be available to subsequent tasks as well.
  * 
- * @param     name     name of the variable to set
+ * @param     name    name of the variable to set
  * @param     val     value to set
+ * @param     secret  whether variable is secret.  optional, defaults to false
  * @returns   void
  */
-export function setVariable(name: string, val: string): void {
+export function setVariable(name: string, val: string, secret?: boolean): void {
+    // once a secret always a secret
+    var key: string = getVariableKey(name);
+    if (variableInfo.hasOwnProperty(key)) {
+        secret = secret || variableInfo[key].secret;
+    }
+
+    // store the value
+    var varValue = val || '';
+    debug('set ' + name + '=' + (secret && varValue ? '********' : varValue));
+    if (secret) {
+        vault.storeSecret('SECRET_' + key, varValue);
+    } else {
+        process.env[key] = varValue;
+    }
+
+    // store the metadata
+    variableInfo[key] = new VariableInfo(name, undefined, secret);
```

Comment:
This is confusion because of the way I implemented it. I think I see a better alternative now.

The point here is to keep track of a list of known variables. We need to maintain this list for several reasons (to distinguish between env vars and job vars; to distinguish between secret vars and public; to know the real variable name and not just the formatted env var name).

All that I really want to record here is variable name, and whether it's a secret. I don't want to run into multi-mastering problems by storing the value (env var stores the correct value for public vars, vault stores the correct value for secret vars).

I think it may improve readability to use a separate class "KnownVariable" to keep track of this info.